File: packages/python/src/a2a_lite/parts.py
"""
Multi-modal parts support (Text, File, Data).

OPTIONAL - Only use if you need files or structured data.
Simple text skills work without any of this.

Example (simple - no parts needed):
    @agent.skill("greet")
    async def greet(name: str) -> str:
        return f"Hello, {name}!"

Example (with files - opt-in):
    from a2a_lite import FilePart, DataPart

    @agent.skill("summarize")
    async def summarize(document: FilePart) -> str:
        content = await document.read_text()
        return summarize(content)
"""

from __future__ import annotations

import base64
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Union


@dataclass
class TextPart:
    """Simple text content."""

    text: str

    def to_a2a(self) -> dict[str, Any]:
        return {"text": self.text}

    @classmethod
    def from_a2a(cls, data: dict) -> TextPart:
        return cls(text=data.get("text", ""))


@dataclass
class FilePart:
    """
    File content - can be bytes or a URI.

    Example:
        @agent.skill("process")
        async def process(file: FilePart) -> str:
            if file.is_uri:
                # Download from URI
                content = await fetch(file.uri)
            else:
                # Use bytes directly
                content = file.data

            return process_content(content)
    """

    name: str
    mime_type: str = "application/octet-stream"
    data: bytes | None = None
    uri: str | None = None

    @property
    def is_uri(self) -> bool:
        return self.uri is not None

    @property
    def is_bytes(self) -> bool:
        return self.data is not None

    async def read_bytes(self) -> bytes:
        """Read file content as bytes."""
        if self.data:
            return self.data
        if self.uri:
            import httpx

            async with httpx.AsyncClient() as client:
                response = await client.get(self.uri)
                response.raise_for_status()
                return response.content
        raise ValueError("FilePart has no data or URI")

    async def read_text(self, encoding: str = "utf-8") -> str:
        """Read file content as text."""
        data = await self.read_bytes()
        return data.decode(encoding)

    def to_a2a(self) -> dict[str, Any]:
        if self.uri:
            return {
                "url": self.uri,
                "filename": self.name,
                "mediaType": self.mime_type,
            }
        else:
            return {
                "raw": base64.b64encode(self.data or b"").decode(),
                "filename": self.name,
                "mediaType": self.mime_type,
            }

    @classmethod
    def from_a2a(cls, data: dict) -> FilePart:
        bytes_data = data.get("raw")
        return cls(
            name=data.get("filename", "unknown"),
            mime_type=data.get("mediaType", "application/octet-stream"),
            data=base64.b64decode(bytes_data) if bytes_data else None,
            uri=data.get("url"),
        )

    @classmethod
    def from_path(cls, path: Union[str, Path], mime_type: str | None = None) -> FilePart:
        """Create FilePart from a local file path."""
        path = Path(path)
        if mime_type is None:
            import mimetypes

            mime_type = mimetypes.guess_type(str(path))[0] or "application/octet-stream"
        return cls(
            name=path.name,
            mime_type=mime_type,
            data=path.read_bytes(),
        )


@dataclass
class DataPart:
    """
    Structured JSON data.

    Example:
        @agent.skill("analyze")
        async def analyze(data: DataPart) -> DataPart:
            result = process(data.data)
            return DataPart(data=result)
    """

    data: dict[str, Any]
    mime_type: str = "application/json"

    def to_a2a(self) -> dict[str, Any]:
        return {
            "data": self.data,
            "mediaType": self.mime_type,
        }

    @classmethod
    def from_a2a(cls, data: dict) -> DataPart:
        return cls(data=data.get("data", {}))


@dataclass
class Artifact:
    """
    Rich output artifact.

    Use when you need more than just text/JSON return.

    Example:
        @agent.skill("generate_report")
        async def generate_report(query: str) -> Artifact:
            return Artifact(
                name="report.pdf",
                parts=[
                    TextPart("Summary: ..."),
                    FilePart.from_path("report.pdf"),
                ],
            )
    """

    name: str | None = None
    description: str | None = None
    parts: list = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)

    def add_text(self, text: str) -> Artifact:
        """Add text to artifact."""
        self.parts.append(TextPart(text=text))
        return self

    def add_file(self, file: FilePart) -> Artifact:
        """Add file to artifact."""
        self.parts.append(file)
        return self

    def add_data(self, data: dict[str, Any]) -> Artifact:
        """Add structured data to artifact."""
        self.parts.append(DataPart(data=data))
        return self

    def to_a2a(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "description": self.description,
            "parts": [p.to_a2a() for p in self.parts],
            "metadata": self.metadata,
        }


# Helper to parse incoming parts
def parse_part(data: dict) -> Union[TextPart, FilePart, DataPart]:
    """Parse an A2A v1.0 part dict into the appropriate Part type.

    Part content is detected by presence of the content key:
    ``text`` / ``data`` / ``raw`` or ``url`` (file parts).
    """
    if "text" in data:
        return TextPart.from_a2a(data)
    elif "data" in data:
        return DataPart.from_a2a(data)
    elif "raw" in data or "url" in data:
        return FilePart.from_a2a(data)
    else:
        # Default to text
        return TextPart(text=str(data))

File: packages/typescript/src/parts.ts
/**
 * Multi-modal parts support (Text, File, Data).
 *
 * Optional - only import if you need files or structured data.
 *
 * Example (simple - no parts needed):
 *   agent.skill("greet", async ({ name }) => `Hello, ${name}!`);
 *
 * Example (with files):
 *   import { FilePart } from 'a2a-lite';
 *
 *   agent.skill("summarize", async ({ doc }: { doc: FilePart }) => {
 *     const content = await doc.readText();
 *     return `Summary: ${content.slice(0, 100)}...`;
 *   });
 */

import { readFileSync } from 'fs';
import { basename } from 'path';
import type { LiteTextPart, LiteFilePart, LiteDataPart, LitePart, Artifact as IArtifact } from './types.js';

/**
 * Create a text part.
 */
export function textPart(text: string): LiteTextPart {
  return { type: 'text', text };
}

/**
 * File part with helper methods.
 */
export class FilePart implements LiteFilePart {
  readonly type = 'file' as const;
  readonly name: string;
  readonly mimeType: string;
  readonly data?: Buffer;
  readonly uri?: string;

  constructor(options: {
    name: string;
    mimeType?: string;
    data?: Buffer | string;
    uri?: string;
  }) {
    this.name = options.name;
    this.mimeType = options.mimeType ?? 'application/octet-stream';
    this.uri = options.uri;

    if (options.data) {
      this.data =
        typeof options.data === 'string' ? Buffer.from(options.data) : options.data;
    }
  }

  /**
   * Create from a local file path.
   */
  static fromPath(path: string, mimeType?: string): FilePart {
    const data = readFileSync(path);
    const name = basename(path);
    const guessedMime = mimeType ?? guessMimeType(name);
    return new FilePart({ name, mimeType: guessedMime, data });
  }

  /**
   * Create from A2A v1.0 wire format.
   * File parts arrive as `{ raw: base64, filename, mediaType }` or `{ url, filename, mediaType }`.
   */
  static fromA2A(data: Record<string, unknown>): FilePart {
    return new FilePart({
      name: (data.filename as string) ?? 'unknown',
      mimeType: (data.mediaType as string) ?? 'application/octet-stream',
      data: data.raw ? Buffer.from(data.raw as string, 'base64') : undefined,
      uri: data.url as string | undefined,
    });
  }

  get isUri(): boolean {
    return this.uri !== undefined;
  }

  get isBytes(): boolean {
    return this.data !== undefined;
  }

  /**
   * Read file content as string.
   */
  async readText(encoding: BufferEncoding = 'utf-8'): Promise<string> {
    const bytes = await this.readBytes();
    return bytes.toString(encoding);
  }

  /**
   * Read file content as bytes.
   */
  async readBytes(): Promise<Buffer> {
    if (this.data) {
      return this.data;
    }

    if (this.uri) {
      const response = await fetch(this.uri);
      const arrayBuffer = await response.arrayBuffer();
      return Buffer.from(arrayBuffer);
    }

    throw new Error('FilePart has no data or URI');
  }

  /**
   * Convert to A2A v1.0 wire format.
   */
  toA2A(): Record<string, unknown> {
    if (this.uri) {
      return {
        url: this.uri,
        filename: this.name,
        mediaType: this.mimeType,
      };
    }

    return {
      raw: this.data?.toString('base64') ?? '',
      filename: this.name,
      mediaType: this.mimeType,
    };
  }
}

/**
 * Structured data part.
 */
export class DataPart implements LiteDataPart {
  readonly type = 'data' as const;
  readonly data: Record<string, unknown>;

  constructor(data: Record<string, unknown>) {
    this.data = data;
  }

  static fromA2A(raw: Record<string, unknown>): DataPart {
    return new DataPart((raw.data as Record<string, unknown>) ?? {});
  }

  toA2A(): Record<string, unknown> {
    return {
      data: this.data,
      mediaType: 'application/json',
    };
  }
}

/**
 * Rich output artifact.
 *
 * Example:
 *   const artifact = new Artifact({ name: "report" })
 *     .addText("Summary here")
 *     .addData({ count: 42 })
 *     .addFile(FilePart.fromPath("data.csv"));
 */
export class Artifact implements IArtifact {
  name?: string;
  description?: string;
  parts: LitePart[] = [];
  metadata?: Record<string, unknown>;

  constructor(options?: {
    name?: string;
    description?: string;
    metadata?: Record<string, unknown>;
  }) {
    this.name = options?.name;
    this.description = options?.description;
    this.metadata = options?.metadata;
  }

  addText(text: string): this {
    this.parts.push(textPart(text));
    return this;
  }

  addFile(file: FilePart): this {
    this.parts.push(file);
    return this;
  }

  addData(data: Record<string, unknown>): this {
    this.parts.push(new DataPart(data));
    return this;
  }

  toA2A(): Record<string, unknown> {
    return {
      name: this.name,
      description: this.description,
      parts: this.parts.map((p) => {
        if ('toA2A' in p && typeof p.toA2A === 'function') {
          return p.toA2A();
        }
        if (p.type === 'text') {
          return { text: p.text };
        }
        return p;
      }),
      metadata: this.metadata,
    };
  }
}

/**
 * Parse an A2A v1.0 wire part.
 *
 * Part content is detected by presence of the content key:
 * `text` / `data` / `raw` or `url` (file parts).
 */
export function parsePart(data: Record<string, unknown>): LitePart {
  if ('text' in data) {
    return { type: 'text', text: (data.text as string) ?? '' };
  }
  if ('data' in data) {
    return DataPart.fromA2A(data);
  }
  if ('raw' in data || 'url' in data) {
    return FilePart.fromA2A(data);
  }
  return { type: 'text', text: String(data) };
}

/**
 * Guess mime type from filename.
 */
function guessMimeType(filename: string): string {
  const ext = filename.split('.').pop()?.toLowerCase();

  const mimeTypes: Record<string, string> = {
    txt: 'text/plain',
    html: 'text/html',
    css: 'text/css',
    js: 'application/javascript',
    ts: 'application/typescript',
    json: 'application/json',
    xml: 'application/xml',
    pdf: 'application/pdf',
    png: 'image/png',
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    gif: 'image/gif',
    svg: 'image/svg+xml',
    csv: 'text/csv',
    md: 'text/markdown',
  };

  return mimeTypes[ext ?? ''] ?? 'application/octet-stream';
}
